torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -0,0 +1,110 @@
+from typing import Any, Callable, List, Optional
+
+import torch
+import torch.distributed as dist
+from torch.distributed.optim import ZeroRedundancyOptimizer
+from torch.distributed.optim.zero_redundancy_optimizer import _OverlapStatus
+from torch.nn.parallel.distributed import DistributedDataParallel
+
+
+def hook_then_zero_step(
+    hook: Callable[[Any, dist.GradBucket], torch.futures.Future],
+    ddp: DistributedDataParallel,
+    zero: ZeroRedundancyOptimizer,
+) -> Callable[[Any, dist.GradBucket], torch.futures.Future]:
+    r"""
+    Modifies the given ``hook`` to additionally perform a partial
+    :class:`ZeroRedundancyOptimizer` :meth:`step` using the gradients in the
+    :class:`DistributedDataParallel` gradient bucket provided by the ``hook``.
+
+    Arguments:
+        hook (Any * dist.GradBucket -> torch.futures.Future): the hook to
+            modify.
+        ddp (DistributedDataParallel): the :class:`DistributedDataParallel`
+            instance to use.
+        zero (ZeroRedundancyOptimizer): the :class:`ZeroRedundancyOptimizer`
+            instance to use.
+
+    Returns:
+        The modified hook.
+    """
+    assert zero._is_functional_optim, "ZeroRedundancyOptimizer must be " \
+        "constructed with a functional optimizer class to use this function"
+
+    def hook_then_zero_fn(
+        state,
+        bucket: dist.GradBucket,
+    ) -> torch.futures.Future:
+        r"""
+        Returns a :class:`Future` that gives a gradient bucket tensor and
+        performs a partial :class:`ZeroRedundancyOptimizer` :meth:`step` using
+        the gradients in that bucket.
+
+        Arguments:
+            state: any state for the hook.
+            bucket (dist.GradBucket): the :class:`DistributedDataParallel`
+                gradient bucket.
+        """
+        fut = hook(state, bucket)
+
+        def zero_step(_) -> List[torch.Tensor]:
```

Comment:
without get the future argument and calling `Future.wait()` how do we guarantee that ZeRO computation starts after the AllReduce is done?

Reply:
I believe that `wait()` is called in the `Reducer` on the future returned by the hook, according to Rohan [here](https://github.com/pytorch/pytorch/pull/61678/files#r674440741). I do not have the code pointer at the moment, but that would explain why we do not need a `wait()` call here. I am not sure if this is accurate since I think that I needed the `wait()` call in the two-stream setting to ensure correctness.

Also, for reference, [Rohan's PR for `hook_then_optim()`](https://github.com/pytorch/pytorch/blob/00f3289d7d4c4ce999a07a6f1e8630fdb1cada38/torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py#L94).

Reply:
I think perhaps this is also due to the semantics of [`then()`](https://pytorch.org/docs/stable/futures.html#torch.futures.Future.then):

> In other words, as long as the callback doesn’t switch streams, it can safely manipulate the result without any additional synchronization.

Reply:
> I think perhaps this is also due to the semantics of then():

I see. Makes sense, this is enforced here:

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/ivalue_inl.h#L649